### PR TITLE
Bugfix/offline dashboard grades

### PR DESCRIPTION
**Pull request by JonathanDuvalV**
"No internet connection" is now displayed instead of "An error occured" for dashboard and grade details when using the app in offline mode

### CODE REVIEW

**Review by apomalyn — APPROVED** [review]
LGTM! Just update your branch :wink:

**Comment by AlexandreMarkus** [comment]
@apomalyn ici il manque l'ajout du bottom snack bar comme discuté dans #151

**Comment by apomalyn** [comment]
> @apomalyn ici il manque l'ajout du bottom snack bar comme discuté dans #151

Tu as completement raison!

**Comment by JonathanDuvalV** [comment]
Here's how it looks with the snackbar:

<details>
  <summary>Schedule</summary>
  Simulator Screen Shot - iPhone 12 Pro Max - 2021-06-28 at 19 07 07
</details>

Also unlike with displaying a toast which can be done anywhere, displaying a snackbar needs a reference to the BuildContext. I've added a check inside the initstate of the 3 main pages that require network connectivity (dashboard, schedule and grades) to display the snackbar if needed. Should I add it to every page or are just those 3 enough?

**Comment by MysticFragilist** [comment]
Nice job! I really like it more as a snackbar than a toast! 😄 

> Here's how it looks with the snackbar:
> 
> Schedule
> https://user-images.githubusercontent.com/45638733/123714866-2630af80-d845-11eb-8da0-56f4e97af22b.png
> Also unlike with displaying a toast which can be done anywhere, displaying a snackbar needs a reference to the BuildContext. I've added a check inside the initstate of the 3 main pages that require network connectivity (dashboard, schedule and grades) to display the snackbar if needed. Should I add it to every page or are just those 3 enough?

Personnally, I would also add it to Grade details, Profile, and ETS page as those pages all required internet connection to function properly.

**Review by MysticFragilist — CHANGES_REQUESTED** [review]
Great job on this PR!!! 😄

**Comment by AlexandreMarkus** [comment]
> Here's how it looks with the snackbar:
> 
> <details>
>   <summary>Schedule</summary>
>  https://user-images.githubusercontent.com/45638733/123714866-2630af80-d845-11eb-8da0-56f4e97af22b.png
> </details>
> 
> Also unlike with displaying a toast which can be done anywhere, displaying a snackbar needs a reference to the BuildContext. I've added a check inside the initstate of the 3 main pages that require network connectivity (dashboard, schedule and grades) to display the snackbar if needed. Should I add it to every page or are just those 3 enough?
> 
> 

I would add the snack bar directly to the navigation bar instead of adding it to the **5** pages

File: lib/core/services/networking_service.dart [hunk 1]
```diff
@@ -1,15 +1,46 @@
...
+
   Future<bool> hasConnectivity() async {
     final connectionStatus = await _connectivity.checkConnectivity();
     return connectionStatus != ConnectivityResult.none;
   }
 
+  Future displayOfflineMode(BuildContext context, AppIntl intl) async {
+    if (!await hasConnectivity()) {
+      if (!_isSnackbarActive && !_isSnackbarDismissed) {
+        final snackBar = SnackBar(
```
[the remark this answers is not in the document]
Comment (AlexandreMarkus) [reply lib/core/services/networking_service.dart:22]: Yes, there is a way to add it directly to the app theme

**Comment by MysticFragilist** [comment]
> I would add the snack bar directly to the navigation bar instead of adding it to the **5** pages

Even better!

**Comment by JonathanDuvalV** [comment]
> > Here's how it looks with the snackbar:
> > 
> >   Schedule
> >  https://user-images.githubusercontent.com/45638733/123714866-2630af80-d845-11eb-8da0-56f4e97af22b.png
> > 
> > Also unlike with displaying a toast which can be done anywhere, displaying a snackbar needs a reference to the BuildContext. I've added a check inside the initstate of the 3 main pages that require network connectivity (dashboard, schedule and grades) to display the snackbar if needed. Should I add it to every page or are just those 3 enough?
> 
> I would add the snack bar directly to the navigation bar instead of adding it to the **5** pages

Good idea!

File: lib/core/services/networking_service.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (JonathanDuvalV) [reply lib/core/services/networking_service.dart:22]: Is this fine? I added it to the app theme just like @AlexandreMarkus said

[the remark this answers is not in the document]
Comment (AlexandreMarkus) [reply lib/core/services/networking_service.dart:22]: Excellent!

File: lib/core/services/networking_service.dart [hunk 2]
```diff
@@ -1,15 +1,46 @@
...
+          content: Text(intl.no_connectivity),
+          duration: const Duration(days: 365),
+          action: SnackBarAction(
+            label: intl.close_no_connectivity_snackbar.toString().toUpperCase(),
```
Comment (apomalyn) [inline lib/core/services/networking_service.dart:26]: Why there is a `toString()` here? `intl.close_no_connectivity_snackbar` is already a string

File: test/ui/widgets/bottom_bar_test.dart [hunk 3]
```diff
@@ -16,16 +16,22 @@ import 'package:notredame/core/constants/router_paths.dart';
...
+import '../../mock/services/networking_service_mock.dart';
+
 NavigationService _navigationService;
+NetworkingServiceMock networkingService;
 
 void main() {
   group('BottomBar - ', () {
     setUp(() {
       _navigationService = setupNavigationServiceMock();
+      networkingService = setupNetworkingServiceMock() as NetworkingServiceMock;
```
Comment (apomalyn) [inline test/ui/widgets/bottom_bar_test.dart:29]: Sometimes you cast the return of `setupNetworkingServiceMock` and sometimes not. Could you change it to unified it?
Also, could you remove the private of `_navigationService` please? :smile:

**Review by apomalyn — CHANGES_REQUESTED** [review]
Very good job :smile:

File: test/ui/widgets/bottom_bar_test.dart (hunk 3 above)
Comment (JonathanDuvalV) [reply test/ui/widgets/bottom_bar_test.dart]: Done! 😄

**Comment by JonathanDuvalV** [comment]
After installing this version of the app on my phone, I realized there's an issue when the snackbar text "jumping" when switching between tabs. Still looking for a fix 😅


https://user-images.githubusercontent.com/45638733/124357881-be36ec00-dbeb-11eb-83e4-7bb92cc83e91.mp4

**Comment by AlexandreMarkus** [comment]
> After installing this version of the app on my phone, I realized there's an issue with the snackbar text "jumping" when switching between tabs. Still looking for a fix 😅
> 
>  video.mp4

@apomalyn @camillebrulotte Is someone with an IOS device able to reproduce the jumping text ?

**Comment by camillebrulotte** [comment]
> @apomalyn @camillebrulotte Is someone with an IOS device able to reproduce the jumping text ?

Yes ! It does the same to me.

**Comment by JonathanDuvalV** [comment]
> > After installing this version of the app on my phone, I realized there's an issue with the snackbar text "jumping" when switching between tabs. Still looking for a fix 😅
> > video.mp4
> 
> @apomalyn @camillebrulotte Is someone with an IOS device able to reproduce the jumping text ?

I looked more into it yesterday and also with the help of @camillebrulotte and we couldn't seem to find a fix. I think that having a text displayed under the navbar like it was originally mentionned in the issue could be a good alternative to the snackbar.

**Comment by AlexandreMarkus** [comment]
> > > After installing this version of the app on my phone, I realized there's an issue with the snackbar text "jumping" when switching between tabs. Still looking for a fix 😅
> > > video.mp4
> > 
> > @apomalyn @camillebrulotte Is someone with an IOS device able to reproduce the jumping text ?
> 
> I looked more into it yesterday and also with the help of @camillebrulotte and we couldn't seem to find a fix. I think that having a text displayed under the navbar like it was originally mentionned in the issue could be a good alternative to the snackbar.

Yes sure! Go ahead with this. You are referring to the YouTube picture right?

**Comment by JonathanDuvalV** [comment]
> Yes sure! Go ahead with this. You are referring to the YouTube picture right?

Yes that's what I'm referring to!

**Comment by AlexandreMarkus** [comment]
> > Yes sure! Go ahead with this. You are referring to the YouTube picture right?
> 
> Yes that's what I'm referring to!

Great, please show us the result when it's done

**Comment by JonathanDuvalV** [comment]
I've updated the snackbar to be text under the navbar instead! I wasn't sure which color to use for the background.

<details>
  <summary>Same as background</summary>


</details>

<details>
  <summary>More of warning color</summary>


</details>

**Comment by MysticFragilist** [comment]
> I've updated the snackbar to be text under the navbar instead! I wasn't sure which color to use for the background.
> 
> Same as background
>  
> More of warning color
>  

I think I would go with the first option! Though I think in dark theme, it should be the same color as the navbar

**Comment by JonathanDuvalV** [comment]
> I think I would go with the first option! Though I think in dark theme, it should be the same color as the navbar

Like this?

**Comment by camillebrulotte** [comment]
> > I think I would go with the first option! Though I think in dark theme, it should be the same color as the navbar
> 
> Like this?
> 

Lookss goooooooooood 👌

**Comment by JonathanDuvalV** [comment]
> > > I think I would go with the first option! Though I think in dark theme, it should be the same color as the navbar
> > 
> > 
> > Like this?
> > 
> 
> Lookss goooooooooood 👌

Alright I'll update it!

File: lib/ui/widgets/base_scaffold.dart [hunk 4]
```diff
@@ -35,25 +39,69 @@ class BaseScaffold extends StatelessWidget {
...
+        bottomNavigationBar: _isOffline ? buildOfflineBar(context) : null,
       );
+
+  Widget buildOfflineBar(BuildContext context) {
+    return Stack(
+      alignment: Alignment.center,
+      children: [
+        Container(
+          color: _isLightTheme(context)
+              ? null
```
Comment (MysticFragilist) [inline lib/ui/widgets/base_scaffold.dart:91]: just to be sure `null` put the background color right?

File: lib/ui/widgets/bottom_bar.dart [hunk 5]
```diff
@@ -28,6 +28,7 @@ class BottomBar extends StatelessWidget {
   Widget build(BuildContext context) {
     return BottomNavigationBar(
       type: BottomNavigationBarType.fixed,
+      elevation: 0,
```
Comment (MysticFragilist) [inline lib/ui/widgets/bottom_bar.dart:31]: Just a question but why is this necessary?

**Review by MysticFragilist — CHANGES_REQUESTED** [review]
Think it covers everything, after the small question LGTM!!

File: lib/ui/widgets/bottom_bar.dart (hunk 5 above)
Comment (JonathanDuvalV) [reply lib/ui/widgets/bottom_bar.dart:31]: For the light theme it creates an ugly seperation under the BottomBar

File: lib/ui/widgets/base_scaffold.dart (hunk 4 above)
Comment (JonathanDuvalV) [reply lib/ui/widgets/base_scaffold.dart:91]: It does yes!

File: lib/l10n/intl_en.arb [hunk 6]
```diff
@@ -214,6 +214,7 @@
   "go_to_dashboard": "Go to Dashboard",
 
   "no_connectivity": "No internet connection",
+  "close_no_connectivity_snackbar": "Dismiss",
```
Comment (AlexandreMarkus) [inline lib/l10n/intl_en.arb:217]: Don't think it is needed

File: lib/l10n/intl_fr.arb [hunk 7]
```diff
@@ -212,6 +212,7 @@
   "go_to_dashboard": "Retour à l'accueil",
 
   "no_connectivity": "Aucune connexion internet",
+  "close_no_connectivity_snackbar": "Fermer",
```
Comment (AlexandreMarkus) [inline lib/l10n/intl_fr.arb:215]: Don't think it is needed

**Review by AlexandreMarkus — CHANGES_REQUESTED** [review]

File: lib/l10n/intl_en.arb (hunk 6 above)
Comment (JonathanDuvalV) [reply lib/l10n/intl_en.arb:217]: True good catch!

**Review by MysticFragilist — APPROVED** [review]
LGTM!!! 🚀

**Review by MysticFragilist — APPROVED** [review]

**Review by AlexandreMarkus — APPROVED** [review]
